src/problems/mod.rs

```diff
@@ -1,100 +1,382 @@
+/// restricted to the \[-1, 1) domain on each dimension:
+///
+/// ```
+/// use std::ops::{Range, RangeInclusive};
+/// use mahf::{Problem, SingleObjective, problems::{VectorProblem, LimitedVectorProblem}};
+///
+/// pub struct Sphere {
+///     pub dim: usize,
+/// }
+///
+/// impl Problem for Sphere {
+///     type Encoding = Vec<f64>;
+///     type Objective = SingleObjective;
+///
+///     fn name(&self) -> &str {
+///         "Sphere"
+///     }
+/// }
+///
+/// impl VectorProblem for Sphere {
+///     type Element = f64;
+///
+///     fn dimension(&self) -> usize {
+///         self.dim
+///     }
+/// }
+///
+/// impl LimitedVectorProblem for Sphere {
+///     fn domain(&self) -> Vec<Range<Self::Element>> {
+///         std::iter::repeat(-1.0..1.0).take(self.dim).collect()
+///     }
+/// }
+/// ```
 pub trait LimitedVectorProblem: VectorProblem {
-    /// Returns the range of the given dimension.
-    fn range(&self, dimension: usize) -> Range<Self::T>;
+    /// Returns the bounds of the search space for each element in the vector encoding.
+    ///
+    /// The bounds specify the range of valid values for each element in the vector.
+    fn domain(&self) -> Vec<Range<Self::Element>>;
 }
 
-/// A [Problem] where one can check for the target.
-pub trait HasKnownTarget {
-    /// Returns whether the target has been reached.
-    fn target_hit(&self, target: SingleObjective) -> bool;
+/// A single-objective optimization problem with an indicator for reaching the optimum.
+///
+/// If the objective value of the optimum is known directly, implement [`KnownOptimumProblem`]
+/// instead, which provides a blanket implementation for this trait.
+///
+/// # Examples
+///
+/// A simple implementation of the real-valued sphere function `f(x) = x^2`, where
+/// the optimum is known to be 0.
+///
+/// Note that a implementation of [`KnownOptimumProblem`] is more sensible in this case because
+/// the value is known directly.
+///
+/// ```
+/// use std::ops::Range;
+/// use mahf::{Problem, SingleObjective, problems::OptimumReachedProblem};
+///
+/// pub struct Sphere {
+///     pub dim: usize,
+/// }
+///
+/// impl Problem for Sphere {
+///     type Encoding = Vec<f64>;
+///     type Objective = SingleObjective;
+///
+///     fn name(&self) -> &str {
+///         "Sphere"
+///     }
+/// }
+///
+/// impl OptimumReachedProblem for Sphere {
+///     fn optimum_reached(&self, objective: SingleObjective) -> bool {
+///         // Approximately 0 is accepted as optimum.
+///         objective.value() < 1e-8
+///     }
+/// }
+///
+/// // A value smaller as 1e-8 counts as optimum.
+/// let sphere = Sphere { dim: 1 };
+/// assert!(sphere.optimum_reached(1e-9.try_into().unwrap()));
+/// ```
+pub trait OptimumReachedProblem: SingleObjectiveProblem {
+    /// Checks whether the objective value has reached the optimum.
+    fn optimum_reached(&self, objective: SingleObjective) -> bool;
 }
```

Comment:
Maybe we can get rid of this in favor of `KnownOptimumProblem`? This used to exist only because COCO was weirdly limited, but now that the Rust bindings expose the optimum, we don't really need this anymore.

Reply:
How likely do you think is another limitation like that for other problems?

Reply:
Very unlikely. Restricting access to the optimum is purely ideological and does not really make sense.

Reply:
Let's remove it, then.